modules/javafx.graphics/src/main/java/com/sun/javafx/font/freetype/PangoGlyphLayout.java

```diff
@@ -126,16 +132,20 @@ public void layout(TextRun run, PGFont font, FontStrike strike, char[] text) {
             OSPango.pango_attr_list_insert(attrList, attr);
         }
 
-        if (str == 0L) {
-            str = OSPango.g_utf16_to_utf8(text);
+        Long str = runUtf8.get(run);
+        if (str == null) {
+            char[] rtext = Arrays.copyOfRange(text, run.getStart(), run.getEnd());
+            str = OSPango.g_utf16_to_utf8(rtext);
+            runUtf8.put(run, str);
             if (check(str, "Failed allocating UTF-8 buffer.", context, desc, attrList)) {
```

Comment:
Did you really mean to store the run in the map before checking if str is null ?

Reply:
I'm not sure what should happen if str is null. Do we want to be able to retry later? In that case, we shouldn't store it in the map. But if it is fatal, and seems to indicate a problem with rtext (e.g. too big?), so in that case it might not be worth retrying.

Reply:
If `str` is null, you would end up retrying it anyway, since `Map::get` will return `null` the next time (if there is a next time). Also, I think the dispose method might then crash trying to free a null pointer. So I think Phil is right and that you should not add the `str` to the `Map` if it is null.

Reply:
I don't think it will be `null` but it will be `0` in which case it is stored in the `Map`
The only reason the str == null is then that we never tried to convert the Java chars to utf8, unless I'm missing a case?

Reply:
Oh, right. It will be a Long 0, not null. If you store it you will still have the problem I mentioned with dispose unless you add back in the `str != 0` check. And you would need a check for `str != 0` in the layout method so that the second time it doesn't treat it as a valid pointer. It might be safer to not store it if 0, which matches the current behavior?

Reply:
That makes sense indeed. For the dispose call, a valid pointer is required.

Reply:
I wasn't completely sure what the idea was, which is why I phrased it as a question :-)